### PR TITLE
fix(foundation): remove readonly from radio button

### PR DESCRIPTION
**Pull request by yinonov**
`readonly` attribute on a radio button violates a WCAG rule. see #6437

there's no evidence `readonly` should be applied on a radio button. it does on radio group however

- [aria-readonly associated roles on MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-readonly#associated_roles) doesn’t list radio
- https://stackoverflow.com/questions/1953017/why-cant-radio-buttons-be-readonly
- https://www.w3schools.com/jsref/dom_obj_radio.asp

as a matter of fact, native radio ignores `readonly` on interaction

https://user-images.githubusercontent.com/15140652/194741456-a37c97e4-3639-4921-a8a3-0e81bb381034.mov

### CODE REVIEW

**Comment by yinonov** [comment]
closes #6437

File: change/@microsoft-fast-foundation-b9885563-e3e8-4c85-a55d-bac04c627fec.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
```
Comment (nicholasrice) [inline change/@microsoft-fast-foundation-b9885563-e3e8-4c85-a55d-bac04c627fec.json:2]:
```suggestion
  "type": "prerelease",
```

File: change/@microsoft-fast-foundation-b9885563-e3e8-4c85-a55d-bac04c627fec.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "remove readonly support from fast radio",
+  "packageName": "@microsoft/fast-foundation",
+  "email": "yinon@hotmail.com",
+  "dependentChangeType": "patch"
```
Comment (nicholasrice) [inline change/@microsoft-fast-foundation-b9885563-e3e8-4c85-a55d-bac04c627fec.json:6]:
```suggestion
  "dependentChangeType": "prerelease"
```

**Review by nicholasrice — CHANGES_REQUESTED** [review]
Thanks for putting this together! I think there is still a little more that needs to be done here. Previously, the radio itself would prevent changing its checked value when readonly. That behavior essentially needs to be moved instead of removed; the checkbox shouldn't get checked when that radiogroup is readonly.

**Comment by yinonov** [comment]
>  the checkbox shouldn't get checked when that radiogroup is readonly.

you mean _**radio**_ shouldn't get checked?

**Comment by nicholasrice** [comment]
> > the checkbox shouldn't get checked when that radiogroup is readonly.
> 
> you mean _**radio**_ shouldn't get checked?

Yes, sorry - *radio*

**Comment by yinonov** [comment]
I ran the site locally on this branch and it seems to already be covered.

correct me if I'm missing anything here @nicholasrice but this is the expected behavior when radio group is set to `readonly` correct? preventing any change to radio-buttons on user interaction

https://user-images.githubusercontent.com/15140652/195021537-0d082bdb-9710-461f-927c-d6e639cb4dd0.mov

**Comment by nicholasrice** [comment]
> I ran the site locally on this branch and it seems to already be covered.
> 
> correct me if I'm missing anything here @nicholasrice but this is the expected behavior when radio group is set to `readonly` correct? preventing any change to radio-buttons on user interaction
> 
>  Screen.Recording.2022-10-11.at.10.17.53.mov

Odd, yes that is my expectation. When I set `readonly` in storybook on your branch it didn't function correctly, but I'll look again; maybe I messed something up.

**Comment by yinonov** [comment]
> I just tried again - setting the `readonly` attribute on the radiogroup custom element does not prevent the user from changing radio checked values in fast-foundation's storybook app. Try to repro there.

you're right. now covered click, keypress and arrow navigation and tabbing into group.

took the liberty to refactor a const to class getter as it repeated in more than 2 places.

https://github.com/yinonov/fast/blob/11ee0292de3667eeadb164ac3a3210694716301c/packages/web-components/fast-foundation/src/radio/radio.ts#L56-L60

File: packages/web-components/fast-foundation/src/radio/radio.ts [hunk 3]
```diff
@@ -66,6 +53,12 @@ export class FASTRadio extends FormAssociatedRadio implements RadioControl {
     @observable
     public defaultSlottedNodes: Node[];
 
+    get radioGroup() {
```
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/radio/radio.ts:56]: Let's make this private so it's not part of the public interface
```suggestion
    private get radioGroup() {
```

**Review by nicholasrice — APPROVED** [review]
Looks good, with one suggestion. Thanks for making this update!

File: packages/web-components/fast-foundation/src/radio/radio.ts (hunk 3 above)
Comment (yinonov) [reply packages/web-components/fast-foundation/src/radio/radio.ts:56]: Thanks, was just trying to debug what API extractor is upset about. Was close

**Comment by yinonov** [comment]
Can this be merged?

**Review by chrisdholt — APPROVED** [review]

**Review by EisenbergEffect — COMMENTED** [review]
Since we removed the tests for the radio itself...do we have an equivalent test for the group?

**Comment by yinonov** [comment]
> Since we removed the tests for the radio itself...do we have an equivalent test for the group?

radio-group had already featured its own `readonly` member.

anything you think should be added considering [this change](packages/web-components/fast-foundation/src/radio-group/radio-group.ts)?

**Comment by EisenbergEffect** [comment]
> > Since we removed the tests for the radio itself...do we have an equivalent test for the group?
> 
> radio-group had already featured its own `readonly` member.
> 
> anything you think should be added considering [this change](https://github.com/microsoft/fast/pull/6438/files#diff-c279a2e5c0a4ca7c3572755239e49e48a9469230d4600bd292a14ed08d87b5a3R242)?

Do we have a test or some other way to ensure that when setting the group to readonly that the individual radios will be readonly? Does the platform handle that for us?

**Comment by yinonov** [comment]
> > > Since we removed the tests for the radio itself...do we have an equivalent test for the group?
> > 
> > 
> > radio-group had already featured its own `readonly` member.
> > anything you think should be added considering [this change](https://github.com/microsoft/fast/pull/6438/files#diff-c279a2e5c0a4ca7c3572755239e49e48a9469230d4600bd292a14ed08d87b5a3R242)?
> 
> Do we have a test or some other way to ensure that when setting the group to readonly that the individual radios will be readonly? Does the platform handle that for us?

How about testing if setting `checked` property to `true` on radio actually functions (when radio-group is set to `readonly`).
Thing is, it's not hardened in the current implementation. setting a `readonly` radio-group AND radio can be bypassed programmatically.

I guess this should be hardened? in the scope of this PR?

https://user-images.githubusercontent.com/15140652/198115081-354eec8b-a6dc-4d05-a676-eed891606e0c.mov

**Comment by chrisdholt** [comment]
@yinonov @olaf-k Given the refactor for "disabled" behavior in PR, etc and the new tests added, I'd like to rebase this and test once those changes are in and follow-up on the tests/behavior at that point. I think the changes made in #6601 will lend themself more towards ensuring that radios are treated as "readonly" from an interactive standpoint. My only other thought there is ensuring that they're communicated in the a11y tree as readonly...but that we can validate at the same time. Thoughts?

**Comment by chrisdholt** [comment]
@yinonov - with the radio/radio group refactor in to solve the state issue it looks like we have some conflicts here. I'd love to get this in. Do we want to resolve conflicts on this branch or just kick off a new one to remove readonly? Happy to move it through quickly regardless of the approach :)

**Comment by yinonov** [comment]
> @yinonov - with the radio/radio group refactor in to solve the state issue it looks like we have some conflicts here. I'd love to get this in. Do we want to resolve conflicts on this branch or just kick off a new one to remove readonly? Happy to move it through quickly regardless of the approach :)

no sweat, resolved the conflicts

**Comment by yinonov** [comment]
Can this be merged?

**Comment by chrisdholt** [comment]
> Can this be merged?

It looks like there are some build failures directly related to the PR. Once we get a passing build I'll pull this down, validate the behavior and merge. Thanks!

image

**Comment by yinonov** [comment]
soz. build works now (locally)

image

**Comment by chrisdholt** [comment]
@yinonov looks like there are still a couple failures - I created a PR against your branch to address these (or you can run prettier and `yarn` locally to address). https://github.com/yinonov/fast/pull/1

Once this is in/updated I'd like to get this in!

**Comment by yinonov** [comment]
Thanks @chrisdholt. Merged

**Pull request merged**